### DISCUSSION TITLE
Reactive transport propblem with a kinetics reaction

### DISCUSSION THREAD

**Discussion by hjunglab** [reaktoro/reaktoro#347] [Q&A]
### Description

Hi Dr. Leal,

Thank you for addressing my previous question. Currently, I am working on a reactive transport problem at the pore scale and exploring Reaktoro to see if I can use Reaktoro as a geochemical reaction solver for my problem. My transport code currently communicates with Reaktoro only through vectors of solutes and solids masses. This limitation requires initiating Reaktoro every single timestep. For now, I would rather focus on coupling Reaktoro without modifying my original code as much as possible.

My problem involves dissolution and precipitation reactions of carbonate minerals, necessitating the use of the KineticsSolver. The mass conservation in my scenario is formulated with chemical elements, meaning that I need to specify the amount of elements rather than species. However, I've encountered a dilemma with the "state.set()" function, which defines the amount of species, and the "conditions.setInitialComponentAmounts()" function, which is only applicable to the equilibrium solver.

I have distilled my questions into two key points:

1. Does the "state.set()" function actually specify the amount of species? For instance, if I use arguments like state.set("Ca+2", 1, mol), it seems like I am specifying the amount of the element ("Ca") rather than the species ("Ca+2") when there are no other sources, such as carbonate minerals. If this is the case, can I use this function to update the amount of elements in a reactive transport code?

2. If I must utilize the "conditions.setInitialComponentAmounts()" function, does this imply that I need to run the equilibrium solver first and then execute the kinetics solver sequentially to update the amount of elements resulting from kinetics reactions?

Thanks so much for your consideration and I would greatly appreciate your help on this matter.

Best,
Heewon

### What is your operating system?

None

**Comment by allanleal** [comment]
Hi @hjunglab , this should be possible: use method `setInitialComponentAmounts` in class `EquilibriumConditions` together with `KineticsSolver`. This issue was reported in discussion #321 and fixed in #323. 

Could you please try that and let me know if you find an issue? If you find, please report with a very minimal reproducible example. This way I can more efficiently investigate the issue and also check if your usage is correct.

Answering your questions:

> Does the "state.set()" function actually specify the amount of species? 

Yes, the `set` method in class `ChemicalState` sets the amount of individual species in the chemical system, and not the amount of an element. 

> does this imply that I need to run the equilibrium solver first and then execute the kinetics solver sequentially to update the amount of elements resulting from kinetics reactions?

If your problem involves chemical kinetics + chemical equilibrium, `KineticsSolver` can deal with this simultaneously, without requiring an equilibrium calculation followed by a kinetic reaction step calculation. Reaktoro solves the entire equilibrium+kinetics problem simultaneously. 

Hope this helps. 

Best,
Allan

**Reply by hjunglab** [reply]
Hi @allanleal, 

Thank you very much for your prompt and helpful response. However, I am still having convergence issues. It appears that `KineticsSolver` is highly sensitive to the component amounts set by `setInitialComponentAmounts`. Even slight adjustments to the component amounts result in a convergence problem with `KineticsSolver`, whereas `EquilibriumSolver` functions properly.

Please refer to the attached code below to replicate the convergence failure. The code contains two vectors, x1 and x2, where the difference of each element is on the order of -14 or less. While 'state1,' defined with x1, converges successfully, 'state2,' defined with x2, does not.

Any assistance you can provide would be greatly appreciated

Best,
Heewon

### Example
~~~py
from reaktoro import * 
import numpy as np

def ratefn(props: ChemicalProps):
    As = 1.525839844e-11
    k0 = 10**-0.3
    aprops = AqueousProps(props)
    Omega = aprops.saturationRatio("Calcite")
    rate = As * k0 * (1 - Omega)
    return rate

# Create an aqueous phase 
all_species = ["H2O(aq)", "H+", "OH-",
              "Ca+2", "Ca(HCO3)+", "CaCl+", "CaOH+",
              "Na+", "Cl-", "NaCl(aq)", "NaOH(aq)",
              "HCO3-", "CO3-2", "CO2(aq)"]

solution = AqueousPhase(all_species).set(ActivityModelPitzer())
mineral = MineralPhase("Calcite")
reaction = MineralReaction("Calcite").setRateModel(ratefn)

db = SupcrtDatabase("supcrtbl")
 
system = ChemicalSystem(db, solution, mineral, reaction)

state1 = ChemicalState(system)
specs1 = EquilibriumSpecs(system)
specs1.temperature()
specs1.pressure()

state2 = ChemicalState(system)
specs2 = EquilibriumSpecs(system)
specs2.temperature()
specs2.pressure()

solver1 = KineticsSolver(specs1)
solver2 = KineticsSolver(specs2)

#compoent vectors
x1=[111.01687012358398, 0.3805750530573242, 56.29016022096396, 8.917223166842402e-05, 9.787537500790597e-05, 0.02061048493702072, 6.21606160514401e-05]
x2=[111.01687012358397, 0.3805750530573241, 56.29016022096395, 8.91722316683774e-05, 9.787537500804966e-05, 0.020610484937020668, 6.216061605146539e-05]
delx=[a - b for a, b in zip(x1, x2)]
print(delx)

cond1 = EquilibriumConditions(specs1)
cond1.temperature(25.0, "celsius")
cond1.pressure(1.0, "atm")
cond1.setInitialComponentAmounts(x1)

cond2 = EquilibriumConditions(specs2)
cond2.temperature(25.0, "celsius")
cond2.pressure(1.0, "atm")
cond2.setInitialComponentAmounts(x2)

result1=solver1.solve(state1, 1, cond1)
assert result1.succeeded(), "reaction 1 calculation Failed."

result2=solver2.solve(state2, 1, cond2)
assert result2.succeeded(), "reaction 2 calculation Failed."
~~~
### What is your operating system?

macOS Ventura 13.4

### Additional Information

_No response_

**Reply by allanleal** [reply]
Hi @hjunglab , the chemical states `state1` and `state2` have 1e-16 moles for all species by default. This is not changed in your two calculations above (which both failed here). Please set some amount for H2O(aq) by adding the following lines of code before the calculations happen:

~~~py
state1.set("H2O(aq)", 1.0, "kg")
state2.set("H2O(aq)", 1.0, "kg")
~~~

This will improve the quality of your initial guess when solving the kinetics problem. 

In the context of reactive transport modeling, you may want to prepare chemical states representing boundary and initial conditions using `EquilibriumSolver`. Consider one `ChemicalState` object for each mesh cell/grid point/DOF - set them with initial condition states you prepared. Now, from one time step to the other, you are using "good" chemical states as initial guesses for the subsequent calculations.

Hope this helps.

-- One last thing: Depending on the choice of time steps (say, a very large one for the current chemical conditions), it's possible that a solution does not exist and the solver will emit the convergence error. In this case, you should try decreasing the time step.

**Reply by hjunglab** [reply]
Hi @allanleal,

Thank you very much for your helpful response, but I am still having a problem.
To provide context, my reactive transport problem focuses on calcite dissolution induced by the injection of an acidic solution. The initial solution was prepared by equilibrating a solution with calcite. So I anticipate no changes in the amount of elements (i.e., no reactions) before introducing an acidic solution. However, the simulation runs only when there is no or a small amount of calcite.  Attempting to include a significant amount of calcite leads to convergence failure, even after providing the necessary chemical states. In addition, the simulation result changes the element amounts which shouldn't be the case. 

I would greatly appreciate it if you could take a closer look at the attached code and offer any suggestions.

Best,
Heewon


~~~py
import numpy as np
from reaktoro import *

def ratefn(props: ChemicalProps):
    As = 1.525839844e-11
    k0 = 10**-0.3
    aprops = AqueousProps(props)
    Omega = aprops.saturationRatio("Calcite")
    rate = As * k0 * (1 - Omega)
    return rate
    
# Create an aqueous phase 
all_species = ["H2O(aq)", "H+", "OH-",
              "Ca+2", "Ca(HCO3)+", "CaCl+", "CaOH+",
              "Na+", "Cl-", "NaCl(aq)", "NaOH(aq)",
              "HCO3-", "CO3-2", "CO2(aq)"]

solution = AqueousPhase(all_species).set(ActivityModelPitzer())
mineral = MineralPhase("Calcite")
reaction = MineralReaction("Calcite").setRateModel(ratefn)

# Create the database
db = SupcrtDatabase("supcrtbl")

# equilibrate the solution with calcite
eq_system = ChemicalSystem(db, solution, mineral)

eq_state = ChemicalState(eq_system)
eq_state.set("H2O(aq)", 1.0,  "kg")
eq_state.set("Na+"    , 2.05, "mg")
eq_state.set("Ca+2"   , 1.42, "mg")
eq_state.set("Cl-"    , 3.47, "mg")
eq_state.set("CO2(aq)", 0.36, "mol")
eq_state.set("Calcite", 10.0,  "mol")

eq_specs = EquilibriumSpecs(eq_system)
eq_specs.temperature()
eq_specs.pressure()

eq_solver = EquilibriumSolver(eq_specs)

eq_cond = EquilibriumConditions(eq_specs)
eq_cond.temperature(25.0, "celsius")
eq_cond.pressure(1.0, "atm")

eq_result=eq_solver.solve(eq_state, eq_cond)
assert eq_result.succeeded(), "Reaktoro equilibrium calculation Failed."

eq_props = eq_state.props()
eq_aqComp=eq_props.componentAmountsInPhase(0).asarray().tolist()


# use the equilibrated water and run a kinetics reaction
kn_system = ChemicalSystem(db, solution, mineral, reaction)

kn_state = ChemicalState(kn_system)
kn_state.set("H2O(aq)", eq_props.speciesAmount("H2O(aq)").val(),  "mol")
kn_state.set("Na+"    , eq_props.speciesAmount("Na+").val(), "mol")
kn_state.set("Ca+2"   , eq_props.speciesAmount("Ca+2").val(), "mol")
kn_state.set("Cl-"    , eq_props.speciesAmount("Cl-").val(), "mol")
kn_state.set("CO2(aq)", eq_props.speciesAmount("CO2(aq)").val(), "mol")
kn_state.set("Calcite", eq_props.speciesAmount("Calcite").val(),  "mol")
# kn_state.set("Calcite", 0.001,  "mol")  # It runs when there is only a small amount of calcite

kn_specs = EquilibriumSpecs(kn_system)
kn_specs.temperature()
kn_specs.pressure()

kn_solver = KineticsSolver(kn_specs)

kn_cond = EquilibriumConditions(kn_specs)
kn_cond.temperature(25.0, "celsius")
kn_cond.pressure(1.0, "atm")
kn_cond.setInitialComponentAmounts(eq_aqComp)

kn_result = kn_solver.solve(kn_state, 0.0001, kn_cond)
assert kn_result.succeeded(), "Reaktoro kinetics calculation Failed."

kn_props = kn_state.props()
kn_aqComp=kn_props.componentAmountsInPhase(0).asarray().tolist()

**Reply by allanleal** [reply]
Hi @hjunglab , please study the code below. It uses only `KineticsSolver` and only one chemical system, with the defined reaction. The first kinetic calculation is performed with zero time step. This enables the aqueous species to equilibrate among themselves, while keeping them in disequilibrium with calcite. The second kinetic calculation uses a modified amounts of elements/components. The script is only meant to show you the "mechanics" of achieving what you want. I could not see calcite dissolving. Maybe worth reviewing the reaction rate and/or initial condition. 

Your previous approach had some issues such as preparing a second chemical state with only amounts of certain species from the first calculation. Ignoring other species present in the initial state in the kinetic state `kn_state` would cause difference in element amounts.

Let me know if you still need further questions.

```py
import numpy as np
from reaktoro import *

def ratefn(props: ChemicalProps):
    As = 1.525839844e-11
    k0 = 10**-0.3
    aprops = AqueousProps(props)
    Omega = aprops.saturationRatio("Calcite")
    rate = As * k0 * (1 - Omega)
    return rate

# Create an aqueous phase
all_species = ["H2O(aq)", "H+", "OH-",
              "Ca+2", "Ca(HCO3)+", "CaCl+", "CaOH+",
              "Na+", "Cl-", "NaCl(aq)", "NaOH(aq)",
              "HCO3-", "CO3-2", "CO2(aq)"]

solution = AqueousPhase(all_species).set(ActivityModelPitzer())
mineral = MineralPhase("Calcite")
reaction = MineralReaction("Calcite").setRateModel(ratefn)

# Create the database
db = SupcrtDatabase("supcrtbl")

# Equilibrate the solution but let it in disequilibrium with calcite by setting time step to zero
system = ChemicalSystem(db, solution, mineral, reaction)

state0 = ChemicalState(system)
state0.set("H2O(aq)", 1.0,  "kg")
state0.set("Na+"    , 2.05, "mg")
state0.set("Ca+2"   , 1.42, "mg")
state0.set("Cl-"    , 3.47, "mg")
state0.set("CO2(aq)", 0.36, "mol")
state0.set("Calcite", 10.0,  "mol")

kn_solver = KineticsSolver(system)

dt = 0.0  # zero time step!

result = kn_solver.solve(state0, dt)

assert result.succeeded(), "Reaktoro equilibrium calculation Failed."

# Print initial state where aqueous species are equilibrated among themselves but in disequilibrium with calcite
print(state0)

c0 = np.array(state0.componentAmounts())

# Perturb c0 as if 0.001 mol of NaCl were added
c0[system.elements().index("Na")] += 0.001
c0[system.elements().index("Cl")] += 0.001

# Perturb c0 as if 0.001 mol of CaCl2 were added
c0[system.elements().index("Ca")] += 0.001
c0[system.elements().index("Cl")] += 0.002

# Perform an kinetic calculation using dt = 1 second with perturbed element amounts
cond = EquilibriumConditions(system)
cond.temperature(25.0, "celsius")
cond.pressure(1.0, "atm")
cond.setInitialComponentAmounts(c0)

dt = 1.0

assert result.succeeded(), "Reaktoro equilibrium calculation Failed."

# Print the resulting state 
print(state0)
```
